core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,48 +147,92 @@ public void setStoreListenerSource(Cache<K, V> source) {
   }
 
   @Override
-  public void onEvent(final CacheEvent<K, V> event) {
+  public void onEvent(CacheEvent<K, V> event) {
     final EventType type = event.getType();
     LOGGER.trace("Cache Event notified for event type {}", type);
-    Map<EventListenerWrapper, Future<?>> notificationResults = 
-        new HashMap<CacheEventNotificationServiceImpl.EventListenerWrapper, Future<?>>(registeredListeners.size());
-    
-    for (final EventListenerWrapper wrapper: registeredListeners) {
-      if (!wrapper.config.fireOn().contains(type)) {
-        continue;
-      }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    eventThreadLocal.get().add(cacheEventWrapper);
+    markFailedCacheEventsInThreadLocal();
+    if (!orderedListenersSet.isEmpty()) {
+      ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue(event.getKey());
+      eventWrapperQueue.add(cacheEventWrapper);
+    }
+  }
+  
+  private void dispatchEvent(CacheEventWrapper cacheEventWrapper) {
+    cacheEventWrapper.markFireable();
+    ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue((K)cacheEventWrapper.cacheEvent.getKey());
+    orderedEventWait(cacheEventWrapper, eventWrapperQueue);
+    if (!orderedListenersSet.isEmpty()) {
+      while (!eventWrapperQueue.isEmpty() && eventWrapperQueue.peek().isFireable()) {
+        CacheEventWrapper event = eventWrapperQueue.peek();
+        actualDispatch(event);
+        cacheEventWrapper.markFired();
+        if (event != eventThreadLocal.get().get(0)) {
+          event.lockEvent();
+          event.firedEventCondition.signal();
```

Comment:
This logic ensures that only the thread that created the event at the front of the `eventQueue` processes `FIREABLE` events, while the other threads wait for their events to get `FIRED`. Thus no matter what order the events become `FIREABLE` (i.e. threads reach the dispatch logic) the event firing order will be maintained. Understandably, this also means that the calling thread that created the event at the front of the `eventQueue` will be the last one to return after event dispatch.